clustering/ejb/hotrod/src/main/java/org/wildfly/clustering/ejb/hotrod/bean/HotRodBeanManager.java

```diff
@@ -0,0 +1,179 @@
+ * @param <V> the bean instance type
+ * @param <M> the bean metadata value type
+ */
+public class HotRodBeanManager<K, V extends BeanInstance<K>, M> implements BeanManager<K, V, TransactionBatch> {
+    private static final Logger LOGGER = Logger.getLogger(BeanManagerNearCacheFactory.class);
+
+    private final RemoteCache<?, ?> cache;
+    private final BeanFactory<K, V, M> beanFactory;
+    private final Supplier<K> identifierFactory;
+    private final BeanExpirationConfiguration<K, V> expiration;
+    private final Batcher<TransactionBatch> batcher;
+    private final Affinity strongAffinity;
+    private final Affinity weakAffinity;
+    private final UnaryOperator<Bean<K, V>> transformer;
+    private final ListenerRegistrar expirationListenerRegistrar;
+
+    private volatile ListenerRegistration expirationListenerRegistration;
+
+    public HotRodBeanManager(HotRodBeanManagerConfiguration<K, V, M> configuration, ListenerRegistrar expirationListenerRegistrar) {
+        this.expirationListenerRegistrar = expirationListenerRegistrar;
+        this.beanFactory = configuration.getBeanFactory();
+        this.identifierFactory = configuration.getIdentifierFactory();
+        this.expiration = configuration.getExpiration();
+        this.cache = configuration.getCache();
+        this.batcher = new HotRodBatcher(this.cache);
+        Group group = configuration.getGroup();
+        this.strongAffinity = new ClusterAffinity(group.getName());
+        RemoteCacheConfiguration cacheConfiguration = this.cache.getRemoteCacheContainer().getConfiguration().remoteCaches().get(this.cache.getName());
+        this.weakAffinity = cacheConfiguration.nearCacheMode().enabled() ? new NodeAffinity(group.getLocalMember().getName()) : Affinity.NONE;
+        // If bean has timeout = 0, remove bean on close
+        Consumer<Bean<K, V>> closeTask = (this.expiration != null) && this.expiration.getTimeout().isZero() ? bean -> bean.remove(this.expiration.getExpirationListener()) : null;
+        this.transformer = (closeTask != null) ? bean -> new OnCloseBean<>(bean, closeTask) : UnaryOperator.identity();
+    }
+
+    @Override
+    public void start() {
+        // We only need to listen for server-side expirations if bean timeout is non-zero.
+        if ((this.expiration != null) && !this.expiration.getTimeout().isZero()) {
+            this.expirationListenerRegistration = this.expirationListenerRegistrar.register();
+        }
+    }
+
+    @Override
+    public void stop() {
+        if (this.expirationListenerRegistration != null) {
+            this.expirationListenerRegistration.close();
+        }
+    }
+
+    @Override
+    public Affinity getStrongAffinity() {
+        return this.strongAffinity;
+    }
+
+    @Override
+    public Affinity getWeakAffinity(K id) {
+        return this.weakAffinity;
+    }
+
+    @Override
+    public int getActiveCount() {
+        return (int) (this.cache.clientStatistics().getNearCacheSize() / 3);
```

Comment:
This method impl deserves a comment.  In summary, this computes an estimate, since HotRod lacks a mechanism to iterate/filter over entries in the local caffeine cache.